### PR TITLE
Suggestion: Add initial support for Lerna monorepos

### PR DESCRIPTION
**Pull request by OriMarron**
This is a suggestion to support searching the `node_modules` of child packages in `lerna` monorepos.

(related to #10)

### background ###
Workspaces that use `lerna` usually have the following file structure:
```
/node_modules
/package.json
/lerna.json

/packages/project-a 
  /node_modules
  /package.json

/packages/project-b/
  /node_modules
  /package.json
```

Currently, for workspaces with this structure, the extension will search only in the root `/node_modules`, leaving the user unable to search in one of the child projects.

With this PR, the user will be prompted to select which project to search in. 
image

### Logic ###
The logic that is used is as follows:

1. If there is no `lerna.json` file at the root level - continue as before
2. read `lerna.json` file, and look at the `packages` field. 
Example json config file:
```
{
  "packages": [
    "packages/*"
  ],
  "version": "0.0.0"
}
```
3. Search for packages matching the pattern(s) using `glob`.
a package is a folder containing a `pakage.json` file

4. If no packages are found - search the root `/node_modules` as before

5. Prompt the user to choose a project to search in: either the root or one of the packages

6. Search the selected project's `/node_modules` folder using the existing logic

Note: patterns containing `**` are ignored for now

Please tell me what you think. 
If you agree with this direction, I would gladly extend the support to also include `**` and `yarn` projects.
@jasonnutter

### CODE REVIEW

**Comment by jasonnutter** [comment]
This looks great, thanks for the contribution! I tried a couple times to get this working and kept getting stuck, nice to see someone get it figured out.

I have some small pieces of feedback, would you like that now, or were you thinking of adding the yarn support to this PR (if you think yarn support may drastically change this PR, I'll wait to provide any detailed feedback until that's ready)?

Thanks!

**Comment by orimwix** [comment]
Please provide any feedback! I will address it as soon as possible :)

**Comment by orimwix** [comment]
I just added a commit with support for `Yarn` workspaces.

It works exactly the same, apart from looking in `package.json/workspaces` instead of `lerna.json/packages`

**Comment by jasonnutter** [comment]
Awesome, I'll take a closer look today! 

One question: one thing I kept getting hung up on was how to properly implement support for when VS Code / yarn / lerna workspaces are nested, e.g. a lerna workspace inside a VS Code workspace. Would that be supported with this implementation?

**Comment by jasonnutter** [comment]
Ultimately not a blocker if not, as that is something that could be added incrementally. Just wanted to see if that was considered.

**Comment by orimwix** [comment]
I actually kept it in mind, and this case is fully supported.

Let's say your workspace looks like this:
```
/regular-repo
    /node_modules
    /package.json

/lerna-repo
    /node_modules
    /lerna.json
    /packages/project1
        /package.json
        /node_modules

    /packages/project2
        /package.json
        /node_modules

```
you will first be prompted to choose between `regular-repo` and `lerna-repo`. If you chose `lerna-repo`, only then you will be prompted to choose between `project1` and `project2`

File: find-child-packages.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    return matches.map(match => path.join(match, '..'));
+};
+
+const getLernaPackagesConfig = async root => {
+    if (!(await exists(path.join(root, LERNA_CONFIG_FILE)))) {
+        return [];
+    }
+
+    const config = await loadJsonFile(path.join(root, LERNA_CONFIG_FILE));
```
Comment (jasonnutter) [inline find-child-packages.js]: If this fails for some reason, e.g. malformed json, maybe we could show an error message?

File: find-child-packages.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    return Array.isArray(config.packages) ? config.packages : [];
+};
+
+const getYarnWorkspacesConfig = async root => {
+    if (!(await exists(path.join(root, PACKAGE_JSON_FILE)))) {
+        return [];
+    }
+
+    const config = await loadJsonFile(path.join(root, PACKAGE_JSON_FILE));
```
Comment (jasonnutter) [inline find-child-packages.js]: Same here about catching errors and showing an error message.

File: find-child-packages.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+const PACKAGE_JSON_FILE = 'package.json';
+const LERNA_CONFIG_FILE = 'lerna.json';
+
+const flat = arrays => [].concat.apply([], arrays);
+
+const distinct = array => [ ...new Set(array) ];
+
+const findPatternMatches = async (root, pattern) => {
+    if (pattern.includes('**')) return [];
```
Comment (jasonnutter) [inline find-child-packages.js]: I like the consts above for lerna/package.json, could we make this a const as well, to make it clear what this string is?

File: find-child-packages.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    if (!(await exists(path.join(root, LERNA_CONFIG_FILE)))) {
+        return [];
+    }
+
+    const config = await loadJsonFile(path.join(root, LERNA_CONFIG_FILE));
+    return Array.isArray(config.packages) ? config.packages : [];
+};
+
+const getYarnWorkspacesConfig = async root => {
+    if (!(await exists(path.join(root, PACKAGE_JSON_FILE)))) {
```
Comment (jasonnutter) [inline find-child-packages.js]: Small nitpick: I personally like seperating these into two statements, makes it a little easier to read.

```js
const packageJsonExists = await exists(path.join(root, PACKAGE_JSON_FILE));
if (!packageJsonExists) {
  // ... 
}
```

**Review by jasonnutter — CHANGES_REQUESTED** [review]
Left a couple comments, otherwise this looks great, thanks again for the contribution!

**Comment by orimwix** [comment]
Thanks for your quick review and feedback!
I committed the requested changes

**Review by jasonnutter — APPROVED** [review]

**Comment by jasonnutter** [comment]
Apologies for the delay, have been on vacation and forgot to circle back on this. I'll get this merged and published this weekend. Thanks again!

**Comment by OriMarron** [comment]
@jasonnutter Hi, any news on publishing a new version?

**Comment by jasonnutter** [comment]
Sorry for the delay, working on it! See my comment on #13

**Comment by OrenZak** [comment]
@jasonnutter when will be a release ?

**Pull request merged**